code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceVolumeLookupPreCreate.java

```diff
@@ -0,0 +1,111 @@
+
+    private static final Logger log = LoggerFactory.getLogger(InstanceVolumeLookupPreCreate.class);
+    private static final String LOCAL = "local";
+
+    @Inject
+    ObjectManager objectManager;
+    @Inject
+    ObjectProcessManager processManager;
+    @Inject
+    StoragePoolDao storagePoolDao;
+    @Inject
+    VolumeDao volumeDao;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "instance.create" };
+    }
+
+    /**
+     * Looks for shared volumes in the dataVolumes field and converts them to dataVolumeMounts.
+     * If volumeDriver is set:
+     *  if set to local, do nothing
+     *  if set to anything else, restrict volume lookup to a matching storage pool
+     *  if blank, do not restrict volume lookup to a particular storage pool.
+     * If volume is found, remove from dataVolumes list and put it in dataVolumeMounts. 
+     * If volume is not found, just leave it in dataVolumes.
+     */
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Instance instance = (Instance)state.getResource();
+        if (!InstanceConstants.KIND_CONTAINER.equals(instance.getKind())) {
+            return null;
+        }
+
+        String vd = DataAccessor.fieldString(instance, InstanceConstants.FIELD_VOLUME_DRIVER);
+        if (LOCAL.equals(vd)) {
+            return null;
+        }
+
+        Long storagePoolId = null;
+        long accountId = instance.getAccountId();
+        if (StringUtils.isNotEmpty(vd)) {
+            StoragePool sp = storagePoolDao.findStoragePoolByDriverName(accountId, vd);
+            if (sp != null) {
+                storagePoolId = sp.getId();
+            }
+        }
+
+        Map<Object, Object> data = new HashMap<>();
+        List<String> dataVolumes = DataAccessor.fieldStringList(instance, InstanceConstants.FIELD_DATA_VOLUMES);
+        Map<String, Object> dataVolumeMounts = DataAccessor.fieldMap(instance, InstanceConstants.FIELD_DATA_VOLUME_MOUNTS);
+        List<String> newDataVolumes = new ArrayList<String>();
+        data.put(InstanceConstants.FIELD_DATA_VOLUMES, newDataVolumes);
+        data.put(InstanceConstants.FIELD_DATA_VOLUME_MOUNTS, dataVolumeMounts);
+        for (String v : dataVolumes) {
+            if (!v.startsWith("/")) {
+                String[] parts = v.split(":", 2);
+                if (parts.length > 1) {
+                    try {
+                        Volume vol = volumeDao.findSharedVolume(accountId, storagePoolId, parts[0]);
+                        if (vol != null) {
+                            dataVolumeMounts.put(parts[1], vol.getId());
+                            continue;
+                        }
+                    } catch (IllegalStateException e) {
+                        log.error(e.getMessage());
+                        objectProcessManager.scheduleProcessInstance(InstanceConstants.PROCESS_REMOVE, instance, null);
```

Comment:
to speed up the removal, we can call scheduleProcessInstanceAsync here. Also curious whether we should handle situation when remove process fails to schedule here. The follow up exception will never be thrown